// chevrotain/lib_esm/src/parse/grammar/gast/gast_visitor_public.js
import { Alternation, Flat, NonTerminal, Option, Repetition, RepetitionMandatory, RepetitionMandatoryWithSeparator, RepetitionWithSeparator, Rule, Terminal } from "./gast_public";
var GAstVisitor = /** @class */ (function () {
    function GAstVisitor() {
    }
    GAstVisitor.prototype.visit = function (node) {
        /* istanbul ignore next */
        if (node instanceof NonTerminal) {
            return this.visitNonTerminal(node);
        }
        else if (node instanceof Flat) {
            return this.visitFlat(node);
        }
        else if (node instanceof Option) {
            return this.visitOption(node);
        }
        else if (node instanceof RepetitionMandatory) {
            return this.visitRepetitionMandatory(node);
        }
        else if (node instanceof RepetitionMandatoryWithSeparator) {
            return this.visitRepetitionMandatoryWithSeparator(node);
        }
        else if (node instanceof RepetitionWithSeparator) {
            return this.visitRepetitionWithSeparator(node);
        }
        else if (node instanceof Repetition) {
            return this.visitRepetition(node);
        }
        else if (node instanceof Alternation) {
            return this.visitAlternation(node);
        }
        else if (node instanceof Terminal) {
            return this.visitTerminal(node);
        }
        else if (node instanceof Rule) {
            return this.visitRule(node);
        }
        else {
            throw Error("non exhaustive match");
        }
    };
    GAstVisitor.prototype.visitNonTerminal = function (node) { };
    GAstVisitor.prototype.visitFlat = function (node) { };
    GAstVisitor.prototype.visitOption = function (node) { };
    GAstVisitor.prototype.visitRepetition = function (node) { };
    GAstVisitor.prototype.visitRepetitionMandatory = function (node) { };
    GAstVisitor.prototype.visitRepetitionMandatoryWithSeparator = function (node) { };
    GAstVisitor.prototype.visitRepetitionWithSeparator = function (node) { };
    GAstVisitor.prototype.visitAlternation = function (node) { };
    GAstVisitor.prototype.visitTerminal = function (node) { };
    GAstVisitor.prototype.visitRule = function (node) { };
    return GAstVisitor;
}());
export { GAstVisitor };
